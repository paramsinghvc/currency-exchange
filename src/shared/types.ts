import { HomeState } from "scenes/Home/home.redux";

export interface IRootState {
  home: HomeState;
}

export interface IObject<T = any> {
  [k: string]: T;
}
