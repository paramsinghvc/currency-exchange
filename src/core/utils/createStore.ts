import { applyMiddleware, createStore } from "redux";
import createSagaMiddleware from "redux-saga";

import rootReducer from "./rootReducer";
import { configureSaga } from "./rootSaga";
import logger from "./loggerMiddleware";

export function getPreviousState() {
  const store = localStorage.getItem("store");
  if (store) {
    try {
      const parsedStore = JSON.parse(store);
      return parsedStore;
    } catch (e) {
      return undefined;
    }
  }
  return undefined;
}

export default (preloadedState = getPreviousState()) => {
  const sagaMiddleware = createSagaMiddleware();
  const middlewares = [logger<any>(), sagaMiddleware];
  const middlewareEnhancer = applyMiddleware(...middlewares);

  const store = createStore(rootReducer, preloadedState, middlewareEnhancer);
  sagaMiddleware.run(configureSaga());
  window.onbeforeunload = function() {
    localStorage.setItem("store", JSON.stringify(store.getState()));
  };
  return store;
};
