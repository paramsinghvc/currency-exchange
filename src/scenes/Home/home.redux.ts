import {
  createActionWithPayload,
  IAction,
  makeConstants,
  createReduxOperation,
  augmentReducer,
  IReduxOperations
} from "@mollycule/redux-operation";
import { Currency } from "shared/models/Currency";

const SET_CURRENCY_MODAL_STATUS = "SET_CURRENCY_MODAL_STATUS";
export const HomeConstants = makeConstants([SET_CURRENCY_MODAL_STATUS]);

const {
  actions: [fetchCurrencyDataRequest, fetchCurrencyDataSuccess, fetchCurrencyDataFailure],
  constants: fetchCurrencyDataConstants,
  reducer: fetchCurrencyDataReducer
} = createReduxOperation("FETCH_CURRENCY_DATA");

export const fetchExchangeRateSaga = createActionWithPayload("FETCH_EXCHANGE_RATE_DATA");

export const setCurrencyModalStatus = createActionWithPayload<symbol, boolean>(
  HomeConstants.get(SET_CURRENCY_MODAL_STATUS) as symbol
);

export const initialState = {
  currencyModalStatus: false
};

const homeReducer = (state = initialState, { type, payload }: IAction<symbol, string>) => {
  switch (type) {
    case HomeConstants.get(SET_CURRENCY_MODAL_STATUS):
      return { ...state, currencyModalStatus: payload };
    default:
      return state;
  }
};

export type HomeState = typeof initialState & { currencyData: IReduxOperations<Currency[]> };

export default augmentReducer<HomeState, IAction<symbol, any>>(homeReducer)({
  currencyData: fetchCurrencyDataReducer
});

export { fetchCurrencyDataRequest, fetchCurrencyDataSuccess, fetchCurrencyDataFailure, fetchCurrencyDataConstants };
