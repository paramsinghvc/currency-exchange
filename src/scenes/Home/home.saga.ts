import { xml2js } from "xml-js";
import { put, delay, takeLatest, call } from "redux-saga/effects";

import safeGet from "shared/utils/safeGet";
import { Currency, XMLAttributeData } from "shared/models/Currency";
import { fetchCurrencyDataRequest, fetchCurrencyDataSuccess, fetchCurrencyDataFailure } from "./home.redux";

export function* fetchExchangeRateData() {
  try {
    yield put(fetchCurrencyDataRequest());
    const xmlResponse = yield call(
      fetch,
      "https://cors-anywhere.herokuapp.com/https://www.ecb.europa.eu/stats/eurofxref/eurofxref-daily.xml"
    );
    const xmlText = yield xmlResponse.text();
    const jsData = xml2js(xmlText, {
      compact: true,
      ignoreDeclaration: true,
      ignoreInstruction: true,
      ignoreComment: true,
      ignoreCdata: true,
      ignoreDoctype: true,
      ignoreText: true
    });
    const xmlAttributeData = safeGet<XMLAttributeData[]>(jsData, "gesmes:Envelope.Cube.Cube.Cube", []);
    const currencyData: Currency[] = Currency.createFromXMLData(xmlAttributeData);
    yield put(fetchCurrencyDataSuccess(currencyData));
  } catch (e) {
    yield put(fetchCurrencyDataFailure(e));
  }
}

export function* watchExchangeRateData() {
  yield takeLatest("FETCH_EXCHANGE_RATE_DATA", function* pollFetchExchangeRateApi() {
    while (true) {
      yield call(fetchExchangeRateData);
      yield delay(20000);
    }
  });
}
