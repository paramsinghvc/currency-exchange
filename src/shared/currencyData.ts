export default {
  ALL: {
    name: "Albania Lek",
    symbol: "&#76;&#101;&#107;"
  },
  AFN: {
    name: "Afghanistan Afghani",
    symbol: "&#1547;"
  },
  ARS: {
    name: "Argentina Peso",
    symbol: "&#36;"
  },
  AWG: {
    name: "Aruba Guilder",
    symbol: "&#402;"
  },
  AUD: {
    name: "Australia Dollar",
    symbol: "&#36;"
  },
  AZN: {
    name: "Azerbaijan New Manat",
    symbol: "&#1084;&#1072;&#1085;"
  },
  BSD: {
    name: "Bahamas Dollar",
    symbol: "&#36;"
  },
  BBD: {
    name: "Barbados Dollar",
    symbol: "&#36;"
  },
  BYR: {
    name: "Belarus Ruble",
    symbol: "&#112;&#46;"
  },
  BZD: {
    name: "Belize Dollar",
    symbol: "&#66;&#90;&#36;"
  },
  BMD: {
    name: "Bermuda Dollar",
    symbol: "&#36;"
  },
  BOB: {
    name: "Bolivia Boliviano",
    symbol: "&#36;&#98;"
  },
  BAM: {
    name: "Bosnia and Herzegovina Convertible Marka",
    symbol: "&#75;&#77;"
  },
  BWP: {
    name: "Botswana Pula",
    symbol: "&#80;"
  },
  BGN: {
    name: "Bulgaria Lev",
    symbol: "&#1083;&#1074;"
  },
  BRL: {
    name: "Brazil Real",
    symbol: "&#82;&#36;"
  },
  BND: {
    name: "Brunei Darussalam Dollar",
    symbol: "&#36;"
  },
  KHR: {
    name: "Cambodia Riel",
    symbol: "&#6107;"
  },
  CAD: {
    name: "Canada Dollar",
    symbol: "&#36;"
  },
  KYD: {
    name: "Cayman Islands Dollar",
    symbol: "&#36;"
  },
  CLP: {
    name: "Chile Peso",
    symbol: "&#36;"
  },
  CNY: {
    name: "China Yuan Renminbi",
    symbol: "&#165;"
  },
  COP: {
    name: "Colombia Peso",
    symbol: "&#36;"
  },
  CRC: {
    name: "Costa Rica Colon",
    symbol: "&#8353;"
  },
  HRK: {
    name: "Croatia Kuna",
    symbol: "&#107;&#110;"
  },
  CUP: {
    name: "Cuba Peso",
    symbol: "&#8369;"
  },
  CZK: {
    name: "Czech Republic Koruna",
    symbol: "&#75;&#269;"
  },
  DKK: {
    name: "Denmark Krone",
    symbol: "&#107;&#114;"
  },
  DOP: {
    name: "Dominican Republic Peso",
    symbol: "&#82;&#68;&#36;"
  },
  XCD: {
    name: "East Caribbean Dollar",
    symbol: "&#36;"
  },
  EGP: {
    name: "Egypt Pound",
    symbol: "&#163;"
  },
  SVC: {
    name: "El Salvador Colon",
    symbol: "&#36;"
  },
  EEK: {
    name: "Estonia Kroon",
    symbol: "&#107;&#114;"
  },
  EUR: {
    name: "Euro",
    symbol: "&#8364;"
  },
  FKP: {
    name: "Falkland Islands (Malvinas) Pound",
    symbol: "&#163;"
  },
  FJD: {
    name: "Fiji Dollar",
    symbol: "&#36;"
  },
  GHC: {
    name: "Ghana Cedis",
    symbol: "&#162;"
  },
  GIP: {
    name: "Gibraltar Pound",
    symbol: "&#163;"
  },
  GTQ: {
    name: "Guatemala Quetzal",
    symbol: "&#81;"
  },
  GGP: {
    name: "Guernsey Pound",
    symbol: "&#163;"
  },
  GYD: {
    name: "Guyana Dollar",
    symbol: "&#36;"
  },
  HNL: {
    name: "Honduras Lempira",
    symbol: "&#76;"
  },
  HKD: {
    name: "Hong Kong Dollar",
    symbol: "&#36;"
  },
  HUF: {
    name: "Hungary Forint",
    symbol: "&#70;&#116;"
  },
  ISK: {
    name: "Iceland Krona",
    symbol: "&#107;&#114;"
  },
  INR: {
    name: "India Rupee",
    symbol: "&#8377;"
  },
  IDR: {
    name: "Indonesia Rupiah",
    symbol: "&#82;&#112;"
  },
  IRR: {
    name: "Iran Rial",
    symbol: "&#65020;"
  },
  IMP: {
    name: "Isle of Man Pound",
    symbol: "&#163;"
  },
  ILS: {
    name: "Israel Shekel",
    symbol: "&#8362;"
  },
  JMD: {
    name: "Jamaica Dollar",
    symbol: "&#74;&#36;"
  },
  JPY: {
    name: "Japan Yen",
    symbol: "&#165;"
  },
  JEP: {
    name: "Jersey Pound",
    symbol: "&#163;"
  },
  KZT: {
    name: "Kazakhstan Tenge",
    symbol: "&#1083;&#1074;"
  },
  KPW: {
    name: "Korea (North) Won",
    symbol: "&#8361;"
  },
  KRW: {
    name: "Korea (South) Won",
    symbol: "&#8361;"
  },
  KGS: {
    name: "Kyrgyzstan Som",
    symbol: "&#1083;&#1074;"
  },
  LAK: {
    name: "Laos Kip",
    symbol: "&#8365;"
  },
  LVL: {
    name: "Latvia Lat",
    symbol: "&#76;&#115;"
  },
  LBP: {
    name: "Lebanon Pound",
    symbol: "&#163;"
  },
  LRD: {
    name: "Liberia Dollar",
    symbol: "&#36;"
  },
  LTL: {
    name: "Lithuania Litas",
    symbol: "&#76;&#116;"
  },
  MKD: {
    name: "Macedonia Denar",
    symbol: "&#1076;&#1077;&#1085;"
  },
  MYR: {
    name: "Malaysia Ringgit",
    symbol: "&#82;&#77;"
  },
  MUR: {
    name: "Mauritius Rupee",
    symbol: "&#8360;"
  },
  MXN: {
    name: "Mexico Peso",
    symbol: "&#36;"
  },
  MNT: {
    name: "Mongolia Tughrik",
    symbol: "&#8366;"
  },
  MZN: {
    name: "Mozambique Metical",
    symbol: "&#77;&#84;"
  },
  NAD: {
    name: "Namibia Dollar",
    symbol: "&#36;"
  },
  NPR: {
    name: "Nepal Rupee",
    symbol: "&#8360;"
  },
  ANG: {
    name: "Netherlands Antilles Guilder",
    symbol: "&#402;"
  },
  NZD: {
    name: "New Zealand Dollar",
    symbol: "&#36;"
  },
  NIO: {
    name: "Nicaragua Cordoba",
    symbol: "&#67;&#36;"
  },
  NGN: {
    name: "Nigeria Naira",
    symbol: "&#8358;"
  },
  NOK: {
    name: "Norway Krone",
    symbol: "&#107;&#114;"
  },
  OMR: {
    name: "Oman Rial",
    symbol: "&#65020;"
  },
  PKR: {
    name: "Pakistan Rupee",
    symbol: "&#8360;"
  },
  PAB: {
    name: "Panama Balboa",
    symbol: "&#66;&#47;&#46;"
  },
  PYG: {
    name: "Paraguay Guarani",
    symbol: "&#71;&#115;"
  },
  PEN: {
    name: "Peru Nuevo Sol",
    symbol: "&#83;&#47;&#46;"
  },
  PHP: {
    name: "Philippines Peso",
    symbol: "&#8369;"
  },
  PLN: {
    name: "Poland Zloty",
    symbol: "&#122;&#322;"
  },
  QAR: {
    name: "Qatar Riyal",
    symbol: "&#65020;"
  },
  RON: {
    name: "Romania New Leu",
    symbol: "&#108;&#101;&#105;"
  },
  RUB: {
    name: "Russia Ruble",
    symbol: "&#1088;&#1091;&#1073;"
  },
  SHP: {
    name: "Saint Helena Pound",
    symbol: "&#163;"
  },
  SAR: {
    name: "Saudi Arabia Riyal",
    symbol: "&#65020;"
  },
  RSD: {
    name: "Serbia Dinar",
    symbol: "&#1044;&#1080;&#1085;&#46;"
  },
  SCR: {
    name: "Seychelles Rupee",
    symbol: "&#8360;"
  },
  SGD: {
    name: "Singapore Dollar",
    symbol: "&#36;"
  },
  SBD: {
    name: "Solomon Islands Dollar",
    symbol: "&#36;"
  },
  SOS: {
    name: "Somalia Shilling",
    symbol: "&#83;"
  },
  ZAR: {
    name: "South Africa Rand",
    symbol: "&#82;"
  },
  LKR: {
    name: "Sri Lanka Rupee",
    symbol: "&#8360;"
  },
  SEK: {
    name: "Sweden Krona",
    symbol: "&#107;&#114;"
  },
  CHF: {
    name: "Switzerland Franc",
    symbol: "&#67;&#72;&#70;"
  },
  SRD: {
    name: "Suriname Dollar",
    symbol: "&#36;"
  },
  SYP: {
    name: "Syria Pound",
    symbol: "&#163;"
  },
  TWD: {
    name: "Taiwan New Dollar",
    symbol: "&#78;&#84;&#36;"
  },
  THB: {
    name: "Thailand Baht",
    symbol: "&#3647;"
  },
  TTD: {
    name: "Trinidad and Tobago Dollar",
    symbol: "&#84;&#84;&#36;"
  },
  TRY: {
    name: "Turkey Lira",
    symbol: null
  },
  TRL: {
    name: "Turkey Lira",
    symbol: "&#8356;"
  },
  TVD: {
    name: "Tuvalu Dollar",
    symbol: "&#36;"
  },
  UAH: {
    name: "Ukraine Hryvna",
    symbol: "&#8372;"
  },
  GBP: {
    name: "United Kingdom Pound",
    symbol: "&#163;"
  },
  USD: {
    name: "United States Dollar",
    symbol: "&#36;"
  },
  UYU: {
    name: "Uruguay Peso",
    symbol: "&#36;&#85;"
  },
  UZS: {
    name: "Uzbekistan Som",
    symbol: "&#1083;&#1074;"
  },
  VEF: {
    name: "Venezuela Bolivar",
    symbol: "&#66;&#115;"
  },
  VND: {
    name: "Viet Nam Dong",
    symbol: "&#8363;"
  },
  YER: {
    name: "Yemen Rial",
    symbol: "&#65020;"
  },
  ZWD: {
    name: "Zimbabwe Dollar",
    symbol: "&#90;&#36;"
  }
};
