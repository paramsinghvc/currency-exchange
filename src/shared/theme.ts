const theme = {
  primaryColor: "#3A539B",
  borderRadius: "5px",
  borderColor: "#F2F2F2",
  subTextColor: "#979797",
  darkGrayTextColor: "#646464",
  highlightColor: "#425eb1"
};

export default theme;
