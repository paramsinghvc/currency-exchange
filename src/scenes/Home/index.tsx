import React, { FC, useState, useEffect, useCallback, ChangeEvent, FormEvent, useMemo } from "react";
import useRedux from "@mollycule/redux-hook";
import Anime from "shared/components/Anime";
import animejs from "animejs";

import safeGet from "shared/utils/safeGet";
import CurrencyChooser from "./components/CurrencyChooser";
import theme from "shared/theme";
import CoinImg from "assets/Coin.svg";
import {
  Holder,
  ActiveExchangeRate,
  Exchanger,
  ExchangeSection,
  CurrencyDropdown,
  CurrencyAbbr,
  CurrencyText,
  StyledCurrencyInput,
  Separator,
  CoinIcon,
  ButtonSection,
  ExchangeButton,
  ActiveExchangeRateHolder,
  Toast,
  StyledCurrencyInputHolder
} from "./styles";
import { fetchExchangeRateSaga, setCurrencyModalStatus } from "./home.redux";
import { IRootState } from "shared/types";
import { IActionFactory, IReduxOperations } from "@mollycule/redux-operation";
import { Currency } from "shared/models/Currency";

const Home: FC = () => {
  const {
    currencyData,
    currencyModalStatus,
    setCurrencyModalStatus: setCurrencyModalStatusAction,
    fetchExchangeRateSaga: fetchExchangeRateSagaAction
  } = useRedux<
    IRootState,
    { currencyModalStatus: boolean; currencyData: IReduxOperations<Currency[]> },
    {
      setCurrencyModalStatus: IActionFactory<symbol, boolean>;
      fetchExchangeRateSaga: IActionFactory<string, unknown>;
    }
  >(
    state => ({
      currencyModalStatus: state.home.currencyModalStatus,
      currencyData: state.home.currencyData
    }),
    { setCurrencyModalStatus, fetchExchangeRateSaga }
  );

  useEffect(() => {
    fetchExchangeRateSagaAction();
  }, [fetchExchangeRateSagaAction]);

  const closeCurrencyChooser = useCallback(() => {
    setCurrencyModalStatusAction(false);
  }, [setCurrencyModalStatusAction]);

  const [currencyChooserData, setCurrencyChooserData] = useState(currencyData);
  const [selectedCurrencies, setSelectedCurrencies] = useState<Array<Currency>>([]);
  const [currencyAmounts, setCurrencyAmounts] = useState<Array<string>>(["15.34", "19.73"]);
  const [selectedSection, setSelectedSection] = useState(0);

  const findCurrencyWithCode = (currencies: Currency[], code: string) =>
    currencies.find((val: Currency) => val.code === code);

  useEffect(() => {
    if (currencyData.payload) {
      const firstDefaultCurrency = findCurrencyWithCode(currencyData.payload, "GBP");
      const secondDefaultCurrency = findCurrencyWithCode(currencyData.payload, "USD");
      if (firstDefaultCurrency && secondDefaultCurrency)
        setSelectedCurrencies([firstDefaultCurrency, secondDefaultCurrency]);
    }
  }, [currencyData]);

  const handleCurrencyClick = useCallback(
    (sectionIndex: number) => () => {
      setCurrencyChooserData({
        ...currencyData,
        payload: currencyData.payload.filter(datum => datum.code !== selectedCurrencies[sectionIndex].code)
      });
      setCurrencyModalStatusAction(true);
      setSelectedSection(sectionIndex);
    },
    [setCurrencyModalStatusAction, currencyData, selectedCurrencies]
  );

  const roundVal = (val: number) => Math.round(val * 100) / 100;

  type GetUpdatedAmountsOptions = {
    sectionIndex: number;
    sectionValue?: string;
    selectedCurrencies: Currency[];
  };

  const getUpdatedCurrencyAmounts = useCallback(
    ({ sectionIndex, sectionValue, selectedCurrencies }: GetUpdatedAmountsOptions) => {
      const result = [...currencyAmounts];
      const otherSectionIndex = sectionIndex ^ 1;
      if (sectionValue !== undefined) result[sectionIndex] = sectionValue;

      const otherCurrencyAmount =
        (+result[sectionIndex] / selectedCurrencies[sectionIndex].rate) * selectedCurrencies[otherSectionIndex].rate;

      result[otherSectionIndex] = "" + roundVal(otherCurrencyAmount);

      return result;
    },
    [currencyAmounts]
  );

  const updateCurrencyAmount = useCallback(
    (sectionIndex: number) => (e: ChangeEvent<HTMLInputElement | HTMLTextAreaElement>) => {
      const result = getUpdatedCurrencyAmounts({ sectionIndex, sectionValue: e.target.value, selectedCurrencies });
      setCurrencyAmounts(result);
    },
    [setCurrencyAmounts, getUpdatedCurrencyAmounts, selectedCurrencies]
  );

  const activeExchangeRateString = useMemo(() => {
    const firstActiveCurrency = selectedCurrencies[0];
    const secondActiveCurrency = selectedCurrencies[1];
    if (firstActiveCurrency && secondActiveCurrency) {
      const convertedSecondVal = roundVal(secondActiveCurrency.rate / firstActiveCurrency.rate);
      return {
        __html: `${firstActiveCurrency.symbol}1 = ${secondActiveCurrency.symbol}${convertedSecondVal}`
      };
    }
    return { __html: "Â£1 = $1.28" };
  }, [selectedCurrencies]);

  const handleSelectedCurrencyChange = useCallback(
    (currency: Currency) => {
      const result = [...selectedCurrencies];
      result[selectedSection] = currency;
      setSelectedCurrencies(result);

      const updatedCurrencyAmounts = getUpdatedCurrencyAmounts({
        sectionIndex: selectedSection,
        selectedCurrencies: result
      });
      setCurrencyAmounts(updatedCurrencyAmounts);
    },
    [selectedCurrencies, selectedSection, getUpdatedCurrencyAmounts, setCurrencyAmounts]
  );

  const [isOpen, setIsOpen] = useState(false);

  useEffect(() => {
    setTimeout(() => {
      setIsOpen(true);
    }, 1000);

    setTimeout(() => {
      setIsOpen(false);
    }, 4000);
  }, []);

  const [showToast, setShowToast] = useState(false);
  const handleExchange = useCallback(() => {
    setShowToast(true);
    setTimeout(() => {
      setShowToast(false);
    }, 2000);
  }, []);

  return (
    <>
      <Holder>
        <ActiveExchangeRateHolder>
          <Anime
            open={!currencyData.pending && currencyData.success}
            duration={1000}
            onEntering={{ translateY: [-20, 0], opacity: [0, 1] }}
          >
            <ActiveExchangeRate dangerouslySetInnerHTML={activeExchangeRateString} />
          </Anime>
        </ActiveExchangeRateHolder>
        <Exchanger>
          <ExchangeSection>
            <Anime
              open={!currencyData.pending && currencyData.success}
              duration={1000}
              onEntering={{ translateX: ["-100%", 0], opacity: [0, 1], delay: animejs.stagger(100) }}
            >
              <StyledCurrencyInputHolder>
                <StyledCurrencyInput
                  paddingsize="large"
                  fullWidth={false}
                  value={currencyAmounts[0]}
                  onChange={updateCurrencyAmount(0)}
                />
              </StyledCurrencyInputHolder>
              <CurrencyDropdown onClick={handleCurrencyClick(0)}>
                <CurrencyAbbr>{safeGet(selectedCurrencies, "0.code", "")}</CurrencyAbbr>
                <CurrencyText>{safeGet(selectedCurrencies, "0.name", "")}</CurrencyText>
              </CurrencyDropdown>
            </Anime>
          </ExchangeSection>
          <Separator>
            <Anime
              open={currencyData.pending}
              duration={500}
              unmountOnExit={false}
              direction="alternate"
              initProps={{ translateY: -29 }}
              onEntering={{ scale: [1, 1.2], loop: true }}
              onExiting={{ loop: false }}
              easing="linear"
            >
              <CoinIcon src={CoinImg} />
            </Anime>
          </Separator>
          <ExchangeSection>
            <Anime
              open={!currencyData.pending && currencyData.success}
              duration={1000}
              onEntering={{ translateX: ["100%", 0], opacity: [0, 1], delay: animejs.stagger(150) }}
            >
              <StyledCurrencyInputHolder>
                <StyledCurrencyInput
                  paddingsize="large"
                  fullWidth={false}
                  value={currencyAmounts[1]}
                  onChange={updateCurrencyAmount(1)}
                />
              </StyledCurrencyInputHolder>
              <CurrencyDropdown onClick={handleCurrencyClick(1)}>
                <CurrencyAbbr>{safeGet(selectedCurrencies, "1.code", "")}</CurrencyAbbr>
                <CurrencyText>{safeGet(selectedCurrencies, "1.name", "")}</CurrencyText>
              </CurrencyDropdown>
            </Anime>
          </ExchangeSection>
        </Exchanger>
        <ButtonSection>
          <ExchangeButton
            onClick={handleExchange}
            label="EXCHANGE"
            color={theme.primaryColor}
            highlightColor={theme.highlightColor}
            fullWidth={false}
            rounded
            size="x-large"
            disabled={currencyData.pending || !currencyData.success}
          />
        </ButtonSection>
      </Holder>
      <CurrencyChooser
        open={currencyModalStatus}
        onClose={closeCurrencyChooser}
        currencyData={currencyChooserData}
        onChange={handleSelectedCurrencyChange}
      />
      <Anime
        open={showToast}
        duration={500}
        onEntering={{ translateY: ["-100%", 0], opacity: [0, 1] }}
        onExiting={{ translateY: "-100%", opacity: 0 }}
      >
        <Toast>Currency Exchanged successfully</Toast>
      </Anime>
    </>
  );
};

export default Home;
