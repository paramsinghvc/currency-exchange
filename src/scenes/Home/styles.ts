import styled from "@emotion/styled";

import TextField from "shared/components/TextField";
import Button from "shared/components/Button";
import theme from "shared/theme";

export const Holder = styled.main`
  display: grid;
  grid-template-rows: minmax(140px, auto) minmax(180px, auto) 170px;
  justify-content: stretch;
`;

export const ActiveExchangeRate = styled.section`
  background: #f2f2f2;
  color: #646464;
  font-size: 12px;
  padding: 15px 20px;
  border-radius: 5px;
  width: fit-content;
  font-weight: 500;
  font-size: 14px;
  line-height: 14px;
  text-align: center;
  align-self: center;
  justify-self: center;
`;

export const Exchanger = styled.section`
  display: flex;
  justify-content: space-evenly;
  align-items: center;
`;

export const ExchangeSection = styled.section`
  display: flex;
  flex-direction: column;
  align-items: flex-end;
`;

export const CurrencyDropdown = styled.div`
  display: flex;
  flex-direction: column;
  align-items: center;
`;

export const CurrencyAbbr = styled.p`
  color: ${theme.primaryColor};
  font-size: 38px;
  text-transform: uppercase;
  margin: 12px auto 12px auto;
`;

export const CurrencyText = styled.p`
  color: ${theme.subTextColor};
  font-size: 9px;
`;

export const StyledTextField = styled(TextField)`
  max-width: 80px;
  font-size: 18px;
`;

export const Separator = styled.div`
  height: 100%;
  width: 1px;
  background: ${theme.borderColor};
  position: relative;
`;

export const CoinIcon = styled.img`
  position: absolute;
  top: 50%;
  left: -29px;
  transform: translateY(-29px);
`;

export const ButtonSection = styled.section`
  display: flex;
  justify-content: center;
  align-items: center;
`;

export const ExchangeButton = styled(Button)`
  font-weight: 500;
  font-size: 12px;
`;
