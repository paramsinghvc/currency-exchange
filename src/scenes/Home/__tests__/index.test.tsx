import React from "react";
import Home from "../index";
import renderer from "react-test-renderer";
import { createStoreContext } from "@mollycule/redux-hook";
import { IRootState } from "shared/types";
import createStore from "core/utils/createStore";
import { BrowserRouter as Router } from "react-router-dom";

const store = createStore();
const { Provider } = createStoreContext<IRootState>();

describe("Should render Home properly", () => {
  it("should match the home snapshot", () => {
    const home = renderer.create(
      <Provider store={store}>
        <Router>
          <Home />
        </Router>
      </Provider>
    );
    expect(home.toJSON()).toMatchSnapshot();
  });
});
