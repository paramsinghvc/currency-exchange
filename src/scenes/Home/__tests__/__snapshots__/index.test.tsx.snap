// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render Home properly should match the home snapshot 1`] = `
<main
  className="css-o8uaij"
>
  <header
    className="css-uf1ume"
  >
    <section
      className="css-ei4gva"
      onClick={[Function]}
      role="button"
    >
      <span />
      <span />
      <span />
    </section>
  </header>
  <main
    className="css-13qacuc"
  >
    <section
      className="css-k68f27"
    />
    <section
      className="css-194lopa"
    >
      <section
        className="css-1oskuwq"
      />
      <div
        className="css-aijkp"
      >
        <img
          className="css-bobwps"
          src="Coin.svg"
        />
      </div>
      <section
        className="css-1oskuwq"
      />
    </section>
    <section
      className="css-1hppjzv"
    >
      <button
        className="css-zgqnjn"
        color="#3A539B"
        disabled={true}
        label="EXCHANGE"
        onClick={[Function]}
        size="x-large"
      >
        EXCHANGE
      </button>
    </section>
  </main>
</main>
`;
